draft-ietf-jsonpath-base.md

```diff
@@ -1176,6 +1181,72 @@ The semantics of regular expressions are as defined in {{-iregexp}}.
 The logical AND, OR, and NOT operators have the normal semantics of Boolean algebra and
 obey its laws (see, for example, {{BOOLEAN-LAWS}}).
 
+##### Function Extensions {#fnex}
+{: unnumbered}
+
+[^unnumbered-bad]
+
+[^unnumbered-bad]: This should not be an unnumbered section, which it is
+    forced to be by the parent section being unnumbered too.
+    The overall structure where the parts of this subsection go to
+    needs to be decided; one part already is in the IANA
+    Considerations in {{iana-fnex}}
+{:source=" -- cabo"}
+
+Beyond the filter expression functionality defined in the preceding
+subsections, JSONPath defines an extension point that can be used to
+add filter expression functionality: "Function Extensions".
+
+A function extension defines a registered name (see {{iana-fnex}}) that
+can be applied to a sequence of zero or more arguments, resulting in a result.
+A function-expression stands for ("returns") a nodelist or a value.
+The arguments are filter expressions that are to be interpreted as
+nodelists or values.
+For each argument and for the result, the function extension defines
+the "kind" of the item, i.e., whether the item is a value or a
+nodelist ("nodes").
```

Comment:
Perhaps rather than the item being  a value, it could be a singular nodelist. This allows chaining the result of a function as an argument of another function with propagation of the empty nodelist. Whether we make a general rule that empty argument nodelists MUST produce empty result nodelists or whether this is a per-function optional behaviour is something we should decide. I quite like the uniformity and predictability of a general rule.

Reply:
Shouldn't the function definition being able to define the "return" even more concretely as a string or number? The latter would be, what is expected with our `length` function.

Reply:
Also, what happens when a function is passed an argument of the wrong kind? Does this produce a syntax error? Is there any coercion to the correct kind? Are there any "soft" runtime errors in which case what does the function return?

Reply:
Let's see which mistakes can be caught at compile time, and whether we can stick with "no runtime errors".